### PR TITLE
Enhance Quiz Functionality and Code Cleanup

### PR DESCRIPTION
**Pull request by aaladaruncc**
This pull request introduces significant enhancements to the quiz functionality, including the ability to fetch grades at the end of the quiz, as well as code cleanup and consistency improvements across various modules. The specific changes made are as follows:

### Key Changes:
- **Fetch Grades at Quiz End**:
  - Added a new method `fetch_grades` in `api/consumers.py` that fetches and calculates the grades of students based on their responses for the quiz.
  - The grades are categorized into buckets: 100, 80, 60, 40, 20, and 0, and the student usernames are appended to the respective buckets.

- **Update Quiz End Logic**:
  - The `end_quiz` method was modified to send the fetched grades as part of the `quiz_ended` message.

- **Database Migration Updates**:
  - Cleaned up foreign key definitions in migrations to follow a consistent inline style without line breaks for better readability.

- **Model Adjustments**:
  - Ensured that all ForeignKey fields in the model definitions utilize a consistent inline format for maintainability in `api/models.py`.

- **Code Consistency Improvements**:
  - Updated various sections across the application to ensure that JSON responses are formatted consistently, especially around status codes, as seen in `api/views/session_views.py` and `api/views/user_views.py`.
  - Applied consistent formatting for message responses to improve readability in case of errors or unauthorized accesses.

### Benefits:
- Enhances user experience by providing detailed feedback on quiz performance.
- Improves code maintainability and readability, making it easier for future developers to work on the codebase.
- Ensures consistency in API responses, which is crucial for front-end integration.

### CODE REVIEW

File: api/consumers.py [hunk 1]
```diff
@@ -181,10 +181,58 @@ def update_quiz_end_time(self):
...
+                grade = "60"
+            elif percentage >= 30:
+                grade = "40"
+            elif percentage >= 10:
+                grade = "20"
+            else:
+                grade = "0"
+
+            # Append student name to the appropriate bucket
+            grade_buckets[grade].append(student.username)
```
Comment (amhajja) [inline api/consumers.py:232]: Here, you're assuming we're having 5 questions; you should probably do:

`grade_buckets[str(round(percentage, 2))].append(student.username)`

Another approach would be to get the number of questions, then create buckets for every possible grade.

**Review by amhajja — CHANGES_REQUESTED** [review]
Left a comment.

**Comment by aaladaruncc** [comment]
> Left a comment.

Resolved the issue, I updated the way the grade buckets are calculated.

File: api/consumers.py [hunk 2]
```diff
@@ -183,10 +184,48 @@ def update_quiz_end_time(self):
...
+                percentage = 0.0
+
+            # Round to two decimal places
+            percentage = round(percentage, 2)
+            percentage_key = f"{percentage}"
+
+            # Append the student's username to the appropriate bucket
+            grade_buckets[percentage_key].append(student.username)
+
+        return grade_buckets
```
Comment (amhajja) [inline api/consumers.py:227]: There might be an issue with this logic now that I'm taking a second look; `total_responses = responses.count()` may not be the number of questions in the quiz; if a student does not submit an answer, there won't be a record in that table.

Total responses should be retrieved from `api_quiz_session_question`, and it should ignore skipped questions, when you get `correct_responses` you should also ignore the questions that were skipped (using `skipped` attribute in `api_quiz_session_question` and not `skipped_question` in `api_user_response`)

Also, how have you tested this? Is there an easy way for the PR reviewer to test this?

**Review by amhajja — CHANGES_REQUESTED** [review]
Found another issue.

**Review by amhajja — CHANGES_REQUESTED** [review]
I like that you added new consumer tests; however, are you sure the test is running in our Django test workflow (I don't believe it is)?

Also, please see the comments I left, and please resolve flake8 errors.

**Comment by aaladaruncc** [comment]
> I like that you added new consumer tests; however, are you sure the test is running in our Django test workflow (I don't believe it is)?
> 
> Also, please see the comments I left, and please resolve flake8 errors.

The tests for consumers aren't part of the django tests flow right now. But you can pull the code and run the following command to run the test: `pytest ./api/tests/consumer_tests.py`. In a future ticket we can integrate the consumer tests in the github workflow.

**Review by amhajja — APPROVED** [review]
This looks great; we should talk about the test that you added during our meeting on Wednesday/Friday.

**Pull request merged**